File: mint_initial.ts
import { ethers } from "hardhat";
import { TweetIPFactory, StoryNFT } from "../typechain-types";
import { http } from "viem";
import { Account, privateKeyToAccount, Address } from "viem/accounts";
import { StoryClient, StoryConfig } from "@story-protocol/core-sdk";

// Update with your deployed factory address
const FACTORY_ADDRESS = "0x8a1885132A04fe94850BEc4f47C55Ca466C7Bb81";

// Story Protocol SDK setup
const privateKey: Address = `0x${process.env.WALLET_PRIVATE_KEY}`;
const account: Account = privateKeyToAccount(privateKey);
const config: StoryConfig = {
  account: account,
  transport: http(process.env.RPC_PROVIDER_URL || "https://aeneid.storyrpc.io"),
  chainId: "aeneid",
};
const storyClient = StoryClient.newClient(config);


async function main() {
  const [signer] = await ethers.getSigners();
  console.log("Using signer:", signer.address);

  // 1) Attach to your deployed factory
  const factory = (await ethers.getContractAt(
    "TweetIPFactory",
    FACTORY_ADDRESS,
    signer
  )) as TweetIPFactory;

  // 2) Create a new collection
  console.log("👉 Creating a new Tweet collection…");
  const createTx = await factory.createTweetCollection(
    "@alice",        // handle
    0,               // mintPrice
    9000,              // maxSupply
    signer.address,  // royaltyReceiver
    500              // royaltyBP (5%)
  );
  await createTx.wait();
  console.log("✅ Collection created");

  // 3) Compute index of the newly‐deployed collection
  const total = await factory.totalCollections();        // bigint
  const lastIndex = total - 1n;                          // bigint subtraction
  const collection = await factory.allCollections(lastIndex);
  console.log("Using collection address:", collection);

  // 4) Mint & register a dummy tweet into *that* collection
  console.log("👉 Minting & registering a dummy tweet…");
  const mintTx = await factory.registerTweetAsset(
    collection,       // ← now correct!
    signer.address,
    "ipfs://QmDummyUri",
    "Alice Example",
    "@alice",
    new Date().toISOString(),
    false,
    0, 0, 0, 0,
    ["#test"],
    ["@bob886"], // have this increment @name_{num} +1 each time.
    "https://example.com/avatar.jpg",
    "https://twitter.com/alice/status/886", // must be unqiue tweet enforced
    "886", // must be unqiue
    "ipfs://QmDummyScreenshot"
  );
  const receipt = await mintTx.wait();
  console.log("✅ Mint + IP-registration in block", receipt.blockNumber);


  // 5) Get the NFT contract and token ID
  const nft = (await ethers.getContractAt(
    "StoryNFT",
    collection,
    signer
  )) as StoryNFT;

  const supply: bigint = await nft.totalSupply();
  const tokenId = (supply - 1n).toString();

  console.log(`Token ID: ${tokenId}`);
  console.log(`🔗 View your NFT here:\nhttps://aeneid.storyscan.io/token/${collection}/instance/${tokenId}`);

  // 6) Get the IP Asset ID from the IP registry
  console.log("👉 Finding IP Asset ID...");
  const ipAssetId = await getIpAssetId(collection, tokenId);
  console.log(`IP Asset ID: ${ipAssetId}`);

  // 7) Create and attach license terms
  await createAndAttachLicenseTerms(ipAssetId);

  // 8) Mint a license token
  await mintLicenseToken(ipAssetId);

}

async function getIpAssetId(nftContract: string, tokenId: string): Promise<string> {
  try {
    // Fixed: Use the correct property names for IpMetadataForWorkflow
    const ipAssetResponse = await storyClient.ipAsset.register({
      nftContract: nftContract as Address,
      tokenId: tokenId,
      ipMetadata: {
        ipMetadataURI: "",
        ipMetadataHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
        nftMetadataURI: "",
        nftMetadataHash: "0x0000000000000000000000000000000000000000000000000000000000000000"
      }
    });
    
    return ipAssetResponse.ipId || "";
  } catch (error) {
    console.error("Error getting IP Asset ID:", error);
    
    // Alternative approach: Calculate IP Asset ID deterministically
    // The IP Asset ID is typically derived from the NFT contract and token ID
    // Fixed: Use ethers v6 syntax instead of utils
    const ipAssetId = ethers.solidityPackedKeccak256(
      ["address", "uint256"],
      [nftContract, tokenId]
    );
    
    console.log("Using calculated IP Asset ID:", ipAssetId);
    return ipAssetId;
  }
}

async function createAndAttachLicenseTerms(ipAssetId: string) {
  console.log("👉 Creating and attaching license terms...");
  
  try {
    // Create commercial remix license terms
    const licenseResponse = await storyClient.license.registerPILTerms({
      defaultMintingFee: 0n,
      currency: '0x1514000000000000000000000000000000000000', // $WIP token
      royaltyPolicy: '0xBe54FB168b3c982b7AaE60dB6CF75Bd8447b390E', // RoyaltyPolicyLAP
      transferable: true,
      expiration: 0n,
      commercialUse: true,
      commercialAttribution: true,
      commercializerChecker: '0x0000000000000000000000000000000000000000',
      commercializerCheckerData: '0x',
      commercialRevShare: 20, // 20% revenue share
      commercialRevCeiling: 0n,
      derivativesAllowed: true,
      derivativesAttribution: true,
      derivativesApproval: false,
      derivativesReciprocal: true,
      derivativeRevCeiling: 0n,
      uri: '',
      // Fixed: Remove waitForTransaction from txOptions
      txOptions: {}
    });

    console.log(`✅ License terms created with ID: ${licenseResponse.licenseTermsId}`);

    // Attach the license terms to the IP Asset
    const attachResponse = await storyClient.license.attachLicenseTerms({
      licenseTermsId: licenseResponse.licenseTermsId!,
      ipId: ipAssetId as Address,
      // Fixed: Remove waitForTransaction from txOptions
      txOptions: {}
    });

    if (attachResponse.success) {
      console.log(`✅ License terms attached to IP Asset at transaction hash ${attachResponse.txHash}`);
    } else {
      console.log(`License terms already attached to this IP Asset.`);
    }

    return licenseResponse.licenseTermsId;
  } catch (error) {
    console.error("Error creating/attaching license terms:", error);
    throw error;
  }
}

async function mintLicenseToken(ipAssetId: string) {
  console.log("👉 Minting license token...");
  
  try {
    // First, we need to get the license terms ID that was attached to this IP
    // For this example, we'll use a common license terms ID
    // You would typically store this from the previous step
    const licenseResponse = await storyClient.license.mintLicenseTokens({
      licenseTermsId: "1", // This should be the actual license terms ID from the previous step
      licensorIpId: ipAssetId as Address,
      receiver: account.address,
      amount: 1,
      maxMintingFee: BigInt(0),
      maxRevenueShare: 100,
      // Fixed: Remove waitForTransaction from txOptions
      txOptions: {}
    });

    console.log(`✅ License token minted at transaction hash ${licenseResponse.txHash}`);
    console.log(`License Token IDs: ${licenseResponse.licenseTokenIds}`);
    
    return licenseResponse.licenseTokenIds;
  } catch (error) {
    console.error("Error minting license token:", error);
    // This might fail if license terms aren't attached yet, which is okay for demo
    console.log("⚠️  License token minting failed - this is expected if license terms weren't properly attached");
  }
}


main().catch((e) => {
  console.error(e);
  process.exitCode = 1;
});

File: mint_initial.py
#!/usr/bin/env python3
"""
mint_initial_part1.py - Create collection and mint first tweet with real Story Protocol licensing

Requirements:
pip install web3 python-dotenv eth-account story_protocol_python_sdk

Usage:
python mint_initial_part1.py
"""

import os
import json
import time
import random
from datetime import datetime
from typing import Dict, Any, List, Optional
from dotenv import load_dotenv
from web3 import Web3
from eth_account import Account
from eth_typing import HexStr
from story_protocol_python_sdk import StoryClient

# Load environment variables
load_dotenv()

# Configuration - matching TypeScript exactly
FACTORY_ADDRESS = "0x8a1885132A04fe94850BEc4f47C55Ca466C7Bb81"
RPC_URL = os.getenv("RPC_PROVIDER_URL", "https://aeneid.storyrpc.io")
PRIVATE_KEY = os.getenv("WALLET_PRIVATE_KEY")

if not PRIVATE_KEY:
    raise ValueError("WALLET_PRIVATE_KEY environment variable is required")

# Initialize Web3
w3 = Web3(Web3.HTTPProvider(RPC_URL))
if not w3.is_connected():
    raise ConnectionError(f"Cannot connect to RPC at {RPC_URL}")

# Setup account
account = Account.from_key(PRIVATE_KEY)
print(f"Using signer: {account.address}")

# Initialize Story Protocol SDK
aeneid_chain_id = 1315
story_client = StoryClient(w3, account, aeneid_chain_id)

# Contract ABIs - matching TypeScript exactly
TWEET_IP_FACTORY_ABI = [
    {
        "inputs": [
            {"internalType": "string", "name": "handle", "type": "string"},
            {"internalType": "uint256", "name": "mintPrice", "type": "uint256"},
            {"internalType": "uint256", "name": "maxSupply", "type": "uint256"},
            {"internalType": "address", "name": "royaltyReceiver", "type": "address"},
            {"internalType": "uint96", "name": "royaltyBP", "type": "uint96"}
        ],
        "name": "createTweetCollection",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "name": "allCollections",
        "outputs": [{"internalType": "address", "name": "", "type": "address"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "totalCollections",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [
            {"internalType": "address", "name": "collection", "type": "address"},
            {"internalType": "address", "name": "to", "type": "address"},
            {"internalType": "string", "name": "uri", "type": "string"},
            {"internalType": "string", "name": "name_", "type": "string"},
            {"internalType": "string", "name": "handle_", "type": "string"},
            {"internalType": "string", "name": "timestamp_", "type": "string"},
            {"internalType": "bool", "name": "verified_", "type": "bool"},
            {"internalType": "uint256", "name": "comments_", "type": "uint256"},
            {"internalType": "uint256", "name": "retweets_", "type": "uint256"},
            {"internalType": "uint256", "name": "likes_", "type": "uint256"},
            {"internalType": "uint256", "name": "analytics_", "type": "uint256"},
            {"internalType": "string[]", "name": "tags_", "type": "string[]"},
            {"internalType": "string[]", "name": "mentions_", "type": "string[]"},
            {"internalType": "string", "name": "profileImage_", "type": "string"},
            {"internalType": "string", "name": "tweetLink_", "type": "string"},
            {"internalType": "string", "name": "tweetId_", "type": "string"},
            {"internalType": "string", "name": "ipfsScreenshot_", "type": "string"}
        ],
        "name": "registerTweetAsset",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    }
]

STORY_NFT_ABI = [
    {
        "inputs": [],
        "name": "totalSupply",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [
            {"internalType": "address", "name": "receiver", "type": "address"}
        ],
        "name": "mintAndRegisterAndCreateTermsAndAttach",
        "outputs": [
            {"internalType": "uint256", "name": "tokenId", "type": "uint256"},
            {"internalType": "address", "name": "ipId", "type": "address"},
            {"internalType": "uint256", "name": "licenseTermsId", "type": "uint256"}
        ],
        "stateMutability": "nonpayable",
        "type": "function"
    }
]

# Initialize factory contract
factory = w3.eth.contract(address=FACTORY_ADDRESS, abi=TWEET_IP_FACTORY_ABI)


def send_transaction(func_call) -> HexStr:
    """Send a transaction and return the hash"""
    # Build transaction
    transaction = func_call.build_transaction({
        'from': account.address,
        'nonce': w3.eth.get_transaction_count(account.address),
        'gas': 0,  # Will estimate
        'gasPrice': w3.eth.gas_price,
    })
    
    # Estimate gas
    try:
        gas_estimate = w3.eth.estimate_gas(transaction)
        transaction['gas'] = int(gas_estimate * 1.2)  # Add 20% buffer
    except Exception as e:
        print(f"Gas estimation failed: {e}")
        transaction['gas'] = 3000000  # Fallback gas limit
    
    # Sign and send
    signed_txn = w3.eth.account.sign_transaction(transaction, PRIVATE_KEY)
    tx_hash = w3.eth.send_raw_transaction(signed_txn.raw_transaction)
    
    return tx_hash.hex()


def wait_for_receipt(tx_hash: str) -> Dict[str, Any]:
    """Wait for transaction receipt"""
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    if receipt['status'] != 1:
        raise Exception(f"Transaction failed: {tx_hash}")
    return receipt


async def main():
    """Main function matching TypeScript structure exactly"""
    
    # 1) Create a new collection
    print("👉 Creating a new Tweet collection...")
    create_tx_hash = send_transaction(
        factory.functions.createTweetCollection(
            "@alice1",        # handle
            0,               # mintPrice
            10,              # maxSupply
            account.address, # royaltyReceiver
            500              # royaltyBP (5%)
        )
    )
    receipt = wait_for_receipt(create_tx_hash)
    print("✅ Collection created")
    
    # 2) Get the latest collection (the one we just created)
    # First get total number of collections
    total_collections = factory.functions.totalCollections().call()
    # Get the latest collection (last index)
    latest_index = total_collections - 1
    collection = factory.functions.allCollections(latest_index).call()
    print(f"Using collection address: {collection}")
    
    # 3) Mint & register a dummy tweet
    print("👉 Minting & registering a dummy tweet...")
    
    # Generate unique values for this run
    timestamp = int(time.time())
    unique_id = f"{timestamp}{random.randint(1000, 9999)}"
    
    mint_tx_hash = send_transaction(
        factory.functions.registerTweetAsset(
            collection,
            account.address,
            "ipfs://QmDummyUri",
            "Alice Example",
            "@bobby",
            datetime.now().isoformat(),
            False,
            0,
            0,
            0,
            0,
            ["#test"],
            [f"@bobby884"],  # THIS MUST CHANGE EACH RUN
            "https://example.com/avatar.jpg",
            f"https://twitter.com/alice/status/884",  # THIS MUST CHANGE EACH RUN
            "884",  # THIS MUST CHANGE EACH RUN
            "ipfs://QmDummyScreenshot"
        )
    )
    mint_receipt = wait_for_receipt(mint_tx_hash)
    print(f"✅ Mint + IP-registration in block {mint_receipt['blockNumber']}")
    
    # 4) Get the NFT contract and token ID
    nft = w3.eth.contract(address=collection, abi=STORY_NFT_ABI)
    
    supply = nft.functions.totalSupply().call()
    token_id = str(supply - 1)
    
    print(f"Token ID: {token_id}")
    print(f"🔗 View your NFT here:")
    print(f"https://aeneid.storyscan.io/token/{collection}/instance/{token_id}")
    
    # 5) Get the IP Asset ID from the IP registry
    print("👉 Finding IP Asset ID...")
    ip_asset_id = await get_ip_asset_id(collection, token_id)
    print(f"IP Asset ID: {ip_asset_id}")
    
    # 6) Create and attach license terms
    license_terms_id = await create_and_attach_license_terms(ip_asset_id)
    
    # 7) Mint a license token
    await mint_license_token(ip_asset_id, license_terms_id)


async def get_ip_asset_id(nft_contract: str, token_id: str) -> str:
    """Get IP Asset ID - matching TypeScript logic exactly"""
    try:
        # Try to register with Story Protocol SDK (matching TypeScript)
        # The Python SDK uses positional arguments, not a dict
        response = story_client.IPAsset.register(
            nft_contract=nft_contract,
            token_id=int(token_id),  # Python SDK expects int, not string
            ip_metadata={
                "ipMetadataURI": "",
                "ipMetadataHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
                "nftMetadataURI": "",
                "nftMetadataHash": "0x0000000000000000000000000000000000000000000000000000000000000000"
            },
            tx_options={}
        )

        print("💬 DEBUG: IPAsset.register raw response:", response)
        if not isinstance(response, dict):
            # if it's an object, show its attributes too
            print("💬 DEBUG: IPAsset.register response.__dict__:", response.__dict__)
        
        # The response should contain the IP ID
        if hasattr(response, 'ip_id'):
            return response.ip_id
        elif isinstance(response, dict) and 'ip_id' in response:
            return response['ip_id']
        else:
            raise Exception("No ip_id returned from registration")
            
    except Exception as e:
        # If the SDK call fails, propagate the error so you catch it upstream
        raise


async def create_and_attach_license_terms(ip_asset_id: str) -> str:
    """Create and attach license terms using Story Protocol SDK"""
    print("👉 Creating and attaching license terms...")
    
    try:
        # Create commercial remix license terms - matching TypeScript exactly
        # Python SDK uses positional arguments instead of dict
        response = story_client.License.register_pil_terms(
            transferable=True,
            royalty_policy="0xBe54FB168b3c982b7AaE60dB6CF75Bd8447b390E",  # RoyaltyPolicyLAP
            default_minting_fee=0,
            expiration=0,
            commercial_use=True,
            commercial_attribution=True,
            commercializer_checker="0x0000000000000000000000000000000000000000",
            commercializer_checker_data="0x",
            commercial_rev_share=20,  # 20% revenue share
            commercial_rev_ceiling=0,
            derivatives_allowed=True,
            derivatives_attribution=True,
            derivatives_approval=False,
            derivatives_reciprocal=True,
            derivative_rev_ceiling=0,
            currency="0x1514000000000000000000000000000000000000",  # $WIP token
            uri="",
            tx_options={
                "from": account.address,
                # also optipnall specify "gas" or "gasPrice" here
            }
        )
        

        # 2) grab the returned tx_hash and wait for it
        license_terms_id = str(response["license_terms_id"])
        tx_hash = response.get("tx_hash") or response.get("txHash")
        if tx_hash is None:
            raise Exception(f"No tx_hash returned for register_pil_terms: {response!r}")

        print(f"⏳ Waiting for license‐terms creation tx {tx_hash} to confirm…")
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f"✅ License terms created on-chain in block {receipt.blockNumber} (id={license_terms_id})")
 
        
        # Attach the license terms to the IP Asset
        attach_response = story_client.License.attach_license_terms(
            ip_id=ip_asset_id,
            license_template="0xBe54FB168b3c982b7AaE60dB6CF75Bd8447b390E",  # PIL template address
            license_terms_id=int(license_terms_id),
            tx_options={"from": account.address}
        )
        
        if hasattr(attach_response, 'success') and attach_response.success:
            print(f"✅ License terms attached to IP Asset at transaction hash {attach_response.txHash}")
        elif isinstance(attach_response, dict) and attach_response.get('success'):
            print(f"✅ License terms attached to IP Asset at transaction hash {attach_response.get('txHash')}")
        else:
            print("License terms already attached to this IP Asset.")
        
        return license_terms_id
        
    except Exception as e:
        print(f"Error creating/attaching license terms: {e}")
        raise


async def mint_license_token(ip_asset_id: str, license_terms_id: str):
    """Mint license token using Story Protocol SDK"""
    print("👉 Minting license token...")
    
    try:
        # Use the actual license terms ID from the previous step
        # If license_terms_id is None, use "1" as fallback (matching TypeScript)
        terms_id = int(license_terms_id) if license_terms_id else 1
        
        # Python SDK uses positional arguments
        response = story_client.License.mint_license_tokens(
            licensor_ip_id=ip_asset_id,
            license_template="0xBe54FB168b3c982b7AaE60dB6CF75Bd8447b390E",  # PIL template
            license_terms_id=terms_id,
            amount=1,
            receiver=account.address,
            royalty_context="0x",
            max_minting_fee=0,
            max_revenue_share=100,
            tx_options={}
        )
        
        if hasattr(response, 'txHash'):
            print(f"✅ License token minted at transaction hash {response.txHash}")
            print(f"License Token IDs: {response.licenseTokenIds if hasattr(response, 'licenseTokenIds') else 'N/A'}")
        elif isinstance(response, dict):
            print(f"✅ License token minted at transaction hash {response.get('txHash')}")
            print(f"License Token IDs: {response.get('licenseTokenIds')}")
        
        return response.get('licenseTokenIds') if isinstance(response, dict) else getattr(response, 'licenseTokenIds', None)
        
    except Exception as e:
        print(f"Error minting license token: {e}")
        # This might fail if license terms aren't attached yet, which is okay for demo
        print("⚠️  License token minting failed - this is expected if license terms weren't properly attached")


async def use_contract_licensing():
    """Alternative: Use the contract's built-in licensing functions"""
    print("👉 Using contract's built-in licensing functions...")
    
    try:
        # Get the first collection
        collection = factory.functions.allCollections(0).call()
        nft = w3.eth.contract(address=collection, abi=STORY_NFT_ABI)
        
        # Use the contract's mintAndRegisterAndCreateTermsAndAttach function
        tx_hash = send_transaction(
            nft.functions.mintAndRegisterAndCreateTermsAndAttach(account.address)
        )
        receipt = wait_for_receipt(tx_hash)
        
        print("✅ Mint + Register + License Terms created in one transaction")
        print(f"Transaction hash: {receipt['transactionHash'].hex()}")
    except Exception as e:
        print(f"Error using contract licensing: {e}")


if __name__ == "__main__":
    import asyncio
    
    try:
        asyncio.run(main())
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        exit(1)